src/storage/s3/storage.c

```diff
@@ -391,32 +389,59 @@ storageS3ListInternal(
                 queryPrefix = strNewFmt("%s%s", strZ(basePrefix), strZ(expressionPrefix));
         }
 
+        // Create query
+        HttpQuery *query = httpQueryNewP();
+
+        // Add the delimiter to not recurse
+        if (!recurse)
+            httpQueryAdd(query, S3_QUERY_DELIMITER_STR, FSLASH_STR);
+
+        // Use list type 2
+        httpQueryAdd(query, S3_QUERY_LIST_TYPE_STR, S3_QUERY_VALUE_LIST_TYPE_2_STR);
+
+        // Don't specify empty prefix because it is the default
+        if (!strEmpty(queryPrefix))
+            httpQueryAdd(query, S3_QUERY_PREFIX_STR, queryPrefix);
+
         // Loop as long as a continuation token returned
+        HttpRequest *request = NULL;
+
         do
         {
             // Use an inner mem context here because we could potentially be retrieving millions of files so it is a good idea to
             // free memory at regular intervals
             MEM_CONTEXT_TEMP_BEGIN()
             {
-                HttpQuery *query = httpQueryNewP();
+                // Get sync response on first request or async response on continuation
+                HttpResponse *response = NULL;
 
-                // Add continuation token from the prior loop if any
-                if (continuationToken != NULL)
-                    httpQueryAdd(query, S3_QUERY_CONTINUATION_TOKEN_STR, continuationToken);
+                if (request == NULL)
+                    response = storageS3RequestP(this, HTTP_VERB_GET_STR, FSLASH_STR, query);
```

Comment:
Same question I had in the Azure storage.c: is `response` a response or a request?

Reply:
It's the same, HttpResponse.

Reply:
same request for comments as in the Azure.c

Reply:
Same as above.